File: check_brackets.dart
import 'dart:io';

void main(List<String> args) {
  if (args.isEmpty) {
    print('Please provide a file path');
    return;
  }

  final file = File(args[0]);
  final content = file.readAsStringSync();

  final stack = <Map<String, dynamic>>[];
  int lineNumber = 1;

  for (int i = 0; i < content.length; i++) {
    final char = content[i];
    if (char == '\n') {
      lineNumber++;
    } else if (char == '(' || char == '[' || char == '{') {
      stack.add({'char': char, 'line': lineNumber});
    } else if (char == ')' || char == ']' || char == '}') {
      if (stack.isEmpty) {
        print(
          'Error: Unexpected ' + char + ' at line ' + lineNumber.toString(),
        );
        return;
      }
      final top = stack.removeLast();
      final topChar = top['char'];
      final topLine = top['line'];

      final matches = {'(': ')', '[': ']', '{': '}'};
      if (matches[topChar] != char) {
        print(
          'Error: Mismatched brackets at line ' +
              lineNumber.toString() +
              '. Expected ' +
              matches[topChar]! +
              ' but got ' +
              char +
              '. Unclosed bracket from line ' +
              topLine.toString(),
        );
        return;
      }
    }
  }

  if (stack.isNotEmpty) {
    print('Error: Unclosed brackets remaining: ' + stack.toString());
  } else {
    print('All brackets match perfectly!');
  }
}

File: check_brackets.py
import sys

def check_brackets(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()
    
    stack = []
    line_number = 1
    for i, char in enumerate(content):
        if char == '\n':
            line_number += 1
        elif char in '([{':
            stack.append((char, line_number))
        elif char in ')]}':
            if not stack:
                print(f"Error: Unexpected '{char}' at line {line_number}")
                return
            top_char, top_line = stack.pop()
            matches = {'(': ')', '[': ']', '{': '}'}
            if matches[top_char] != char:
                print(f"Error: Mismatched brackets at line {line_number}. Expected '{matches[top_char]}' but got '{char}'. Unclosed bracket from line {top_line}")
                return
    
    if stack:
        print(f"Error: Unclosed brackets remaining: {stack}")
    else:
        print("All brackets match perfectly!")

if __name__ == '__main__':
    check_brackets(sys.argv[1])
